checker/src/call_visitor.rs

```diff
@@ -1190,49 +1188,72 @@ impl<'call, 'block, 'analysis, 'compilation, 'tcx, E>
     #[logfn_inputs(TRACE)]
     fn handle_has_tag(&mut self, checking_presence: bool) {
         precondition!(self.actual_args.len() == 1);
-        let result: Rc<AbstractValue>;
-        if let Some(tag) = self.extract_tag_kind_and_mask() {
+
+        let result: Option<Rc<AbstractValue>>;
+        if let Some(tag) = self.extract_tag_kind_and_propagation_set() {
             let source_path = Path::new_deref(self.actual_args[0].0.clone())
                 .refine_paths(&self.block_visitor.bv.current_environment);
-            debug!(
+            trace!(
                 "MiraiHasTag: checking if {:?} has {}been attached with <{:?},{:?}>",
                 source_path,
                 (if checking_presence { "" } else { "never " }),
                 tag.0,
                 tag.1
             );
 
-            // Obtain the possible tags associated at the source.
-            let source_rustc_type = self
-                .block_visitor
-                .bv
-                .type_visitor
-                .get_path_rustc_type(&source_path, self.block_visitor.bv.current_span);
-            let source_abs = self
-                .block_visitor
-                .bv
-                .lookup_path_and_refine_result(source_path, source_rustc_type);
-            let source_tags = source_abs.get_cached_tags();
+            // Lookup the value located at the source path and check its associated tags.
+            // Also lookup the ancestors of the source path to find possible tags.
+            let mut cur_path = source_path;
+            let mut cur_result = Rc::new(abstract_value::FALSE);
+            loop {
```

Comment:
In other parts of MIRAI the pattern is to loop through all of the (path, value) pairs in the current environment, filtering for paths that are rooted in source_path. You can return from the loop as soon as one of the qualifying (path, value) pairs has value with the tag. Use iterators.

Reply:
I also tried this approach: when adding tags, loop through all of the (path, value) pairs to update the values. An issue for this approach is that some of the paths may not be in the current environment, e.g., struct fields of a function parameter.

Reply:
This is where things get complicated. Essentially, all decisions must be postponed until the right environment is present. Let's discuss this over VC.